oak_functions/wasm/src/lib.rs

```diff
@@ -36,388 +36,470 @@ use oak_functions_abi::{
+        self.logger.log_sensitive(
+            Level::Error,
+            &format!("Unable to read buffer from guest memory: {:?}", err),
+        ); */
+        OakStatus::ErrInvalidArgs
+    })?;
+    Ok(target)
 }
 
-fn check_export_function_signature(
-    instance: &wasmi::ModuleInstance,
-    export_name: &str,
-    expected_signature: &wasmi::Signature,
-) -> anyhow::Result<()> {
-    let export_function = instance
-        .export_by_name(export_name)
-        .context("couldn't find Wasm export")?
-        .as_func()
-        .cloned()
-        .context("couldn't interpret Wasm export as function")?;
-    if export_function.signature() != expected_signature {
-        anyhow::bail!(
-            "invalid signature for export: {:?}, expected: {:?}",
-            export_function.signature(),
-            expected_signature
+///  Writes the u32 `value` at the `address` of the Wasm memory.
+pub fn write_u32(
+    ctx: &mut impl AsContextMut,
+    memory: &mut wasmi::Memory,
+    value: u32,
+    address: AbiPointer,
+) -> Result<(), OakStatus> {
+    let value_bytes = &mut [0; 4];
+    LittleEndian::write_u32(value_bytes, value);
+    write_buffer(ctx, memory, value_bytes, address).map_err(|_err| {
+        // TODO(mschett): Fix logging
+        /*
+        self.logger.log_sensitive(
+            Level::Error,
+            &format!("Unable to write u32 value into guest memory: {:?}", err),
         );
-    } else {
-        Ok(())
+        */
+        OakStatus::ErrInvalidArgs
+    })
+}
+
+/// Writes the buffer `source` at the address `dest` of the Wasm memory, if `source` fits in the
+/// allocated memory.
+pub fn write_buffer(
+    ctx: &mut impl AsContextMut,
+    memory: &mut wasmi::Memory,
+    source: &[u8],
+    dest: AbiPointer,
+) -> Result<(), OakStatus> {
+    // TODO(mschett): Check whether we want to validate range.
+    // self.validate_range(dest, source.len() as u32)?;
+
+    let dest = usize::try_from(dest)
+        .expect("failed to convert AbiPointer to usize as required by wasmi API");
+    memory.write(ctx, dest, source).map_err(|_err| {
+        // TODO(mschett): Add logging.
+        /*
+        self.logger.log_sensitive(
+            Level::Error,
+            &format!("Unable to write buffer into guest memory: {:?}", err),
+        );  */
+        OakStatus::ErrInvalidArgs
+    })
+}
+
+// Calls given alloc Func with ctx and length as parameters.
+// alloc Func has to belong to given ctx.
+pub fn call_alloc(ctx: &mut impl AsContextMut, alloc: Func, len: i32) -> AbiPointer {
+    let inputs = &[wasmi::Value::I32(len)];
+    // TODO(mschett): Check whether putting default value 0 is a good idea.
+    let mut outputs = [wasmi::Value::I32(0); 1];
+    alloc
+        .call(ctx, inputs, &mut outputs)
+        .expect("`alloc` call failed");
+
+    let result_value = outputs;
+
+    match result_value[0] {
+        wasmi::Value::I32(v) => v as u32,
+        _ => panic!("invalid value type returned from `alloc`"),
```

Comment:
We should definitely not panic here, since a malformed / malicious module may trigger this, which may be observable by the host. (this may well be a pre-existing issue, I haven't checked)

Reply:
Done.